File: csharp/44_-_Wildcard_Matching.cs
/**
 * 44. Wildcard Matching - Hard
 * Given an input string (s) and a pattern (p), implement wildcard pattern matching with support for '?' and '*' where:
 * '?' Matches any single character.
 * '*' Matches any sequence of characters (including the empty sequence).
 * The matching should cover the entire input string (not partial).
 */

public class Solution {
    public bool IsMatch(string s, string p) {
        int[][] memo = new int[s.Length][];
        
        for (int i = 0; i < memo.Length; ++i) {
            memo[i] = new int[p.Length];
        }
        
        return DFSMemo(s, p, 0, 0, memo);
    }
    
    public bool DFSMemo(string s, string p, int i, int j, int[][] memo) {
        if (i == s.Length && j == p.Length) {
            return true;
        }
        
        if (j >= p.Length) {
            return false;
        }
        
        if (i == s.Length) {
            for (int k = j; k < p.Length; ++k) {
                if (p[k] != '*') {
                    return false;
                }
            }
            
            return true;
        }
        
        // memo[i][j] values can be
        // 0=unchecked, 1=match, 2=no match
        if (memo[i][j] != 0) {
            return memo[i][j] == 1;
        }
        
        bool matched = false;
        
        if (p[j] == '*') {
            // i + 1, j = greedy matching
            matched = DFSMemo(s, p, i + 1, j, memo);
            
            // i, j + 1 = 0-length matching
            matched = matched || DFSMemo(s, p, i, j + 1, memo);
        } else {
            if (s[i] == p[j] || p[j] == '?') {
                matched = true;
            }
            
            matched = matched && DFSMemo(s, p, i + 1, j + 1, memo);
        }
        
        memo[i][j] = (matched ? 1 : 2);
        return memo[i][j] == 1;
    }
}

File: javascript/44_-_Wildcard_Matching.js
/**
 * 44. Wildcard Matching - Hard
 * Given an input string (s) and a pattern (p), implement wildcard pattern matching with support for '?' and '*' where:
 * '?' Matches any single character.
 * '*' Matches any sequence of characters (including the empty sequence).
 * The matching should cover the entire input string (not partial).
 */

/**
 * @param {string} s
 * @param {string} p
 * @return {boolean}
 */
var isMatch = function(s, p) {
    return dfsMemo(s, p, 0, 0, []);
};

var dfsMemo = function(s, p, i, j, memo) {
    if (i === s.length && j === p.length) {
        return true;
    }
    
    // reached the end of s
    // only valid characters in p that can match are '*' (representing 0-length match)
    if (i === s.length) {
        for (let k = j; k < p.length; ++k) {
            if (p.charAt(k) !== '*') {
                return false;
            }
        }
        
        return true;
    }
    
    // haven't reached the end of the input string
    // but have no pattern characters to match against
    if (j >= p.length) {
        return false;
    }
    
    // memoization - return what we computed already
    if (memo[i] !== undefined && memo[i][j] !== undefined) {
        return memo[i][j];
    }
    
    let matched = false;
    
    if (p.charAt(j) !== '*') {
        if (s.charAt(i) === p.charAt(j) || p.charAt(j) === '?') {
            matched = true;
        }
        
        matched = matched && dfsMemo(s, p, i + 1, j + 1, memo);
    } else {
        // i + 1, j = greedy matching
        matched = dfsMemo(s, p, i + 1, j, memo);
        
        // i, j + 1 = 0-length matching
        matched = matched || dfsMemo(s, p, i, j + 1, memo);
    }
    
    if (memo[i] === undefined) {
        memo[i] = [];
    }
    
    memo[i][j] = matched;
    return memo[i][j];
};